### ISSUE TITLE
Named dictionary inputs and outputs for tf.keras.Model

### ISSUE THREAD

**Issue by huyng** [tensorflow/tensorflow#34114]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

**feature request:**

If the user passes in dictionaries for the `inputs` or `outputs` argument during `tf.keras.Model` initialization, preserve the dictionary datatype in `tf.keras.Model.inputs` and `tf.keras.Model.outputs` (i.e. avoid normalizing these arguments to lists). 

This would allow us to refer to inputs and outputs by name rather than by indices in a list.

**current behavior/background:**

`tf.keras.Model` allows us to construct a `Model` with `inputs` and `outputs` arguments that are passed in as dictionaries of tensors. This is a very nice feature because it  allows you to evaluate a model by passing in dictionaries of named data, and to refer to outputs using their named outputs.

However, if one wants to *extend* a model by attaching more layers to a given output, we lose all of the naming information given to us during the model initialization. This is because the `inputs` and `outputs` arguments are always "flattened" to lists and assigned to `Model.outputs` and `Model.inputs` regardless if the user passed in a dictionary or not for these arguments.

As a result of losing this naming information, we have to fall back to using integer based indexing to retrieve the output tensor of interest that we want to extend. While not the end of the world, this is more error prone since layer ordering may move around due to changes in the initial model.

In the following example, I illustrate how defining, and extending a model with new layers currently works:

```python
from tensorflow.keras import Model
from tensorflow.keras.layers import Input, Dense

def create_base_model():
    # create initial model
    a = Input([10], dtype=tf.float32)
    b = Dense(4)(a)
    c = Dense(8)(b)
    model = Model(inputs={'a': a}, outputs={'b': b, 'c': c})
    return model

def extend_base_model(model):
    # create a new model to attach additional layers to tensor "c"
    c = model.outputs[-2]
    d = Dense(16)(c)
    
    # notice here, since model.inputs and model.outputs are flattened to a lists,
    # we lose all naming information
    new_model = Model(inputs=model.inputs, outputs=model.outputs+[d])
    return new_model
    
base = create_base_model()
model = extend_base_model(base)

# we lose all naming information in the inputs and outputs given by the base model
# outputs: [<tensor>, <tensor> ,  <tensor>]
model(tf.zeros([1, 10], dtype=tf.float32))
```

If the feature request was adopted, here's how defining, extending, and executing models could work:

```python
from tensorflow.keras import Model
from tensorflow.keras.layers import Input, Dense

def create_base_model():
    # create initial model
    a = Input([10], dtype=tf.float32)
    b = Dense(4)(a)
    c = Dense(8)(b)
    model = Model(inputs={'a': a}, outputs={'b': b, 'c': c})
    return model

def extend_base_model(model):
    # create a new model to attach additional layers to tensor "c"
    c = model.outputs["c"]
    d = Dense(16)(c)
    
    new_model = Model(inputs=model.inputs, outputs={'d': d, **model.outputs})
    return new_model
    
base = create_base_model()
model = extend_base_model(base)

# we preserve all naming information in the inputs and outputs given by the base model
# outputs {"b": <tensor>, "c": <tensor> , "d": <tensor>}
model({"a": tf.zeros([1, 10], dtype=tf.float32))
```

**Will this change the current api? How?**

Yes, it will change `Model.input` and `Model.output` data types *if* the user passes in a dictionary for these arguments during initialization. 

current behavior:

```python
a = Input([10], dtype=tf.float32)
b = Dense(4)(a)
model = Model(inputs={'a': a}, outputs={'b': b})
type(model.inputs)  # list
type(model.outputs) # list
```

new behavior:

```python
a = Input([10], dtype=tf.float32)
b = Dense(4)(a)
model = Model(inputs={'a': a}, outputs={'b': b})
type(model.inputs)  # dict
type(model.outputs) # dict
```

**Who will benefit with this feature?**

People designing and developing models will be able to more easily and clearly expose output tensors for downstream users to access. 

Downstream users who need to extend a pre-defined models will more easily be able to access tensors using their named outputs.

**Any Other info.**

Not sure if this is the right way to implement this, but it looks like the `_nested_outputs` and `_nested_inputs` variables listed below preserve the data types of the inputs and outputs arguments that were initially passed into the Model constructor.

https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/keras/engine/network.py#L264-L265

**Comment by lyonguyen8697** [comment]
You can access originally passed `inputs`, `outputs` argments in `Model.input` and `Model.output` (not `Model.inputs`, `Model.outputs`).

**Comment by tanzhenyu** [comment]
This seems a good feature request to have, for now all we support is through DenseFeatures

**Comment by jjedele** [comment]
I just debugged the weirdest effects for more than a day until I found out that some of my outputs are swapped from the order I created them in (which makes sense given it's a dictionary, but duh...).

Definitely +1 for this request and the current behaviour should be better documented.

E.g. what happens when I train my multi-output model with a tf.data.Dataset yielding structured tuples? Are they matched based on their name or also just flattened in some way?

**Comment by jjedele** [comment]
And for other people having this problem:

The models seem to have a `output_names` property which can be used to map the list back to the named outputs.

**Comment by qlzh727** [comment]
Thanks for the suggestion. Since this feature is affecting the existing API contract (type of the return value), I am not sure if we can easily change it in TF2 without breaking other users. Having said that, I think the proposed API is more intuitive, and we probably should consider it.

Adding @fchollet and @omalleyt12 here.

**Comment by omalleyt12** [comment]
There's actually two ways to pass dicts through the Functional API at the moment (which is a little ambiguous, so maybe we should clean this up). IMO here is the most supported way:

```python
a = tf.keras.Input(10, name='a'). # 'a' becomes a key in the dict.
b = tf.keras.Input(10, name='b')
out = tf.keras.layers.Add()([a, b])
model = tf.keras.Model([a, b], [out])

# The keys of the dict are the names of the Inputs.
x = {'a': ..., 'b': ...}. 
y_pred = model(x)
```

Using this pattern, you can then extend the base model like this:

```python
out = model.outputs[0]
out = tf.keras.layers.Dense(1)(out)
extended_model = tf.keras.Model(model.inputs, [out])

# This still works
x = {'a': ..., 'b': ...}. 
y_pred = extended_model(x)
```

This should work in the latest nightly

Since we do allow the other way (directly passing nested structures to the Functional API `tf.keras.Model` constructor), maybe we should change `Model.inputs` and `Model.outputs` to match this structure in that case

Alternatively, we could just drop support for nested structures in the `tf.keras.Model` constructor and require the names of the `tf.keras.Input`s to be passed when wanting to pass a dict through the Functional API, since having two ways to do this is confusing. This would be backwards incompatible though, so we'd have to make sure it's not impacting a large number of users

**Comment by huyng** [comment]
@qlzh727 @omalleyt12  thanks for the detailed response. I was not aware that adding a  `name` parameter  to  `Input` would change the input type for `model()` to a dict. 

While this addresses `Model.inputs`, the naming information is still loss for `Model.outputs` if for example, we define a `Model` that outputs a dictionary:

```python
import tensorflow as tf
c = tf.keras.Input(2, name='c')                          
d = tf.keras.Input(2, name='d')                          
z = tf.keras.layers.Add()([c*10, d])                     
j = tf.keras.layers.Add()([z, z])                        
model = tf.keras.Model(inputs={'c':c, 'd': d}, outputs={'z': z, 'j': j})
```

To extend the model above, we would have to resort back to indexing the `model.outputs` list:

```python
out = model.outputs[0]
k = tf.keras.layers.add()([out, out, out])
extended_model = tf.keras.Model(inputs=model.inputs, outputs=model.outputs+[k])
```
It's not the end of the world that we're losing the naming information in `extended_models`, but it would make defining models so much less error prone if I could rely on the dictionary keys to refer to both `inputs` and `outputs` when extending models. 

Perhaps if backwards incompatible changes are not allowed at this point for tensorflow, would you consider adding a different model definition API that preserves `inputs` and `outputs` types *by default*? E.g. introduce a class called  `Module` where `Module.inputs` and `Module.outputs` is a dictionary if the user passes in a dictionary of tensors and is a list if the user passes in a list of tensors? 

`keras.Model` could even extend its default behavior to present to end users the current API behavior, without causing any backward incompatable changes.

Thanks again for review this issue,
-Huy

---
Also, unrelated to the feature request but I just tried adding the `name` parameter to `Input` and there seems to be some odd behavior with that method:

```python
import tensorflow as tf

c = tf.keras.Input(2, name='c')
d = tf.keras.Input(2, name='d')
out = tf.keras.layers.Add()([c*10, d])
model = tf.keras.Model([c, d], [out])

model({'c': tf.ones([1, 2]) * 2, 'd': tf.ones([1, 2]) * 3}).numpy() 
# As expected, this outputs: array([[23,  23]])

model({'c': tf.ones([1, 2]) * 2, 'a': tf.ones([1, 2]) * 3}).numpy()
# Unexpectedly, this outputs: array([[32, 32]])
# I would expect that this call throws an exception
```

**Comment by huyng** [comment]
Whoops, also just saw @jjedele 's comments regarding `output_names`.  Although this would allow us to reconstruct model.output's naming information, IMO the API proposed above feels a little more intuitive and would be nicer to work with since we wouldn't have to go through the extra trouble of rebuilding the input and output dictionaries every time we want to extend a model.

**Comment by tanzhenyu** [comment]
IIUC this breaks backward compatibility if users were passing a dictionary before? Seems niche case we can drop for the broader benefit here...

**Comment by qlzh727** [comment]
@omalleyt12, is any of your recent updates to model training logic addressing this issue?

**Comment by wileeam** [comment]
@qlzh727 As for the output I think [this](#34691) solves it.

**Comment by nitinmnsn** [comment]
any update on this?

**Comment by ConverJens** [comment]
@huyng @jjedele Regarding the output names: if one persists the Keras model as a SavedModel the output names are lost and are not retrievable. This happens if you use tf serving since then you do not have the option to load the model as a Keras model, only a SavedModel where this information is lost. It would be very nice the output names of the layers are persisted since this greatly facilitates multi output models in a live industry setting.

**Comment by oholimoli** [comment]
Yes please, this feature would be highly appreciated. I have a Model with 4 inputs and 8 outputs, adding and removing outputs is a big pain atm. Especially during development when vizualizing the the results is key for further progress.

**Comment by tanzhenyu** [comment]
I think this is supported from 2.2? Can anyone check how this works:
```
keras_input = tf.keras.Input(shape=(5,))
output = tf.keras.layers.Dense(units=1)(keras_input)
func_model = tf.keras.Model({'a': keras_input}, {'b': output})
func_model({'a': np.random.random((32, 5))})
```

**Comment by piyueh** [comment]
> I think this is supported since 2.2? Can anyone check how this works:
> 
> ```
> keras_input = tf.keras.Input(shape=(5,))
> output = tf.keras.layers.Dense(units=1)(keras_input)
> func_model = tf.keras.Model({'a': keras_input}, {'b': output})
> func_model({'a': np.random.random((32, 5))})
> ```

This piece of code works in v2.3.1, but I think it's just luck because `func_model({'zzzzzzz': np.random.random((32, 5))})` also works in the last line.

**Comment by sushreebarsa** [comment]
@huyng It looks like you are using an older Version of Tensorflow `2.0`. Many bugs have been fixed in the latest version. Can you please execute your code using Latest Version `2.6.0` and let us know if the issue still persists? Please find a similar[ issue ](https://github.com/tensorflow/tensorflow/issues/34691) & let us know if it helps ?Thanks!
